Improve yum module perfomance

MiLk opened the pull request:
The result of the `what_provides` function is filtered to use only the most recent version of each package.
When a package has a huge list of versions available, it will only test if the most recent is installed.

I tested with the following task:

```
- name: "Update yum and plugins"
  yum-devel:
    name={{ item }}
    state=latest
    enablerepo=epel-yum-rawhide
  with_items:
    - yum
    - yum-plugin-fastestmirror
```

Repo file here: http://repos.fedorapeople.org/repos/james/yum-rawhide/epel-yum-rawhide.repo

Before this change without repoquery: 3m57s
After this change without repoquery: 34s

Before this change with repoquery: 4m20s
After this change with repoquery: 20s

Commit:
Improve yum module perfomance

Make the `what_provides` function to return only the most recent version of each package.
When a package have a huge list of versions available, it will only test if the most recent is installed.

mpdehaan commented:
I'm a little wary of the regex in the middle of this code making this non-intuitive to read.   Can you add some comments to this?  (And possibly consider using string operations if possible instead?)

MiLk commented:
I use the regex to extract the name of the package.
The format is `%{name}-%{version}-%{release}.%{arch}`.
I'll try to find a better way to do that.

Commit:
Improve yum latest performance

Don't get anymore the list of the packages providing the spec.
Query only for the latest available version of the package.
If none is returned, then the package is already up-to-date.

MiLk commented:
The new implementation seems to be better.
It's faster too.

It doesn't use anymore `what_provides`.
It queries only the latest available version of the package.

MiLk commented:
Any news?

mpdehaan commented:
Hi!

Thanks very much for your interest in Ansible.  It sincerely means a lot to us. 

On September 26, 2014, due to enormous levels of contribution to the project Ansible decided to reorganize module repos, making it easier
for developers to work on the project and for us to more easily manage new contributions and tickets.

We split modules from the main project off into two repos, http://github.com/ansible/ansible-modules-core and http://github.com/ansible/ansible-modules-extras

If you still would like this pull request merged, we will need your help making this target the new repo.  If you do not take any action, this
pull request unfortunately cannot be applied.

We apologize that we are not able to make this transition happen seamlessly, though this is a one-time change and your help is greatly appreciated -- 
this will greatly improve velocity going forward.

Both sets of modules will ship with Ansible, though they'll receive slightly different ticket handling.

To locate where a module lives between 'core' and 'extras'
- Find the module at http://docs.ansible.com/list_of_all_modules.html
- Open the documentation page for that module
- If the bottom of the docs say "This is an extras module", submit your ticket to https://github.com/ansible/ansible-modules-extras
- Otherwise, submit your pull request to update the existing module to https://github.com/ansible/ansible-modules-core
- Note that python modules in ansible now also end in ".py" and this extension is required for new contributions.
- action_plugins (modules with server side components) still live in the main repo.  If your pull request touches both, which should be
   exceedingly rare, submit two new pull requests and make sure to mention the links to each other in the comments.

Otherwise, if this is a new module:
- Submit your pull request to add a module to https://github.com/ansible/ansible-modules-extras

It may be possible to re-patriate your pull requests automatically, one user-submitted approach for advanced git users
has been suggested at https://gist.github.com/willthames/afbaaab0c9681ed45619

Additionally, should you need more help with this, you can ask questions on:
- the development mailing list: https://groups.google.com/forum/#!forum/ansible-devel

Thank you very much!
